cpp/meta_module/algorithm/meta.cpp

```diff
@@ -0,0 +1,330 @@
+#include "meta.hpp"
+
+#include <atomic>
+#include <mutex>
+
+#include "mgp.hpp"
+
+namespace Meta {
+
+namespace {
+
+struct Metadata {
+  Metadata() : node_cnt{0}, relationship_cnt{0} {}
+
+  void Reset();
+  void UpdateLabels(const mgp::Node &node, int add, bool online);
+  void UpdatePropertyKeyCnt(const auto &node_or_relationship, int add, bool online);
+  void UpdateRelationshipTypes(const mgp::Relationship &relationship, int add, bool online);
+  void UpdateRelationshipTypesCnt(const mgp::Relationship &relationship, int add, bool online);
+
+  std::atomic<int64_t> node_cnt;
+  std::atomic<int64_t> relationship_cnt;
+
+  std::unordered_map<std::string, int64_t> labels;
+  std::unordered_map<std::string, int64_t> property_key_cnt;
+  std::unordered_map<std::string, int64_t> relationship_types;
+  std::unordered_map<std::string, int64_t> relationship_types_cnt;
+};
+
+struct MetadataMutex {
+  std::mutex labels_mutex;
+  std::mutex property_key_cnt_mutex;
+  std::mutex relationship_types_mutex;
+  std::mutex relationship_types_cnt_mutex;
+};
+
+// Global variable
+Metadata metadata;
+MetadataMutex metadata_mutex;
+
+void LockAndExecute(std::mutex &mutex, bool online, auto Function) {
```

Comment:
I think a better name for this is `MaybeLockAndExecute`, because you are maybe locking. Also, naming of `Function` should be `function` (lowercase)